diff --git a/Makefile b/Makefile
index b262c0a..e3b1392 100644
--- a/Makefile
+++ b/Makefile
@@ -145,6 +145,7 @@ UPROGS=\
 	$U/_logstress\
 	$U/_forphan\
 	$U/_dorphan\
+	$U/_test_scheduler\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
@@ -166,7 +167,7 @@ QEMUGDB = $(shell if $(QEMU) -help | grep -q '^-gdb'; \
 	then echo "-gdb tcp::$(GDBPORT)"; \
 	else echo "-s -p $(GDBPORT)"; fi)
 ifndef CPUS
-CPUS := 3
+CPUS := 1
 endif
 
 QEMUOPTS = -machine virt -bios none -kernel $K/kernel -m 128M -smp $(CPUS) -nographic
diff --git a/kernel/proc.c b/kernel/proc.c
index 22a5401..776d75c 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -6,6 +6,29 @@
 #include "proc.h"
 #include "defs.h"
 
+unsigned long rand_next = 1;
+
+int
+do_rand(unsigned long *ctx)
+{
+  long hi, lo, x;
+  x = (*ctx % 0x7ffffffe) + 1;
+  hi = x / 127773;
+  lo = x % 127773;
+  x = 16807 * lo - 2836 * hi;
+  if (x < 0)
+    x += 0x7fffffff;
+  x--;
+  *ctx = x;
+  return x;
+}
+
+int
+rand(void)
+{
+  return do_rand(&rand_next);
+}
+
 struct cpu cpus[NCPU];
 
 struct proc proc[NPROC];
@@ -145,6 +168,8 @@ found:
   memset(&p->context, 0, sizeof(p->context));
   p->context.ra = (uint64)forkret;
   p->context.sp = p->kstack + PGSIZE;
+  p->tickets = 10;
+  p->rounds = 0;
 
   return p;
 }
@@ -226,6 +251,9 @@ userinit(void)
   
   p->cwd = namei("/");
 
+  p->tickets = 10;
+  p->rounds = 0;
+
   p->state = RUNNABLE;
 
   release(&p->lock);
@@ -278,6 +306,8 @@ kfork(void)
 
   // copy saved user registers.
   *(np->trapframe) = *(p->trapframe);
+  np->tickets = p->tickets; 
+  np->rounds = 0;           
 
   // Cause fork to return 0 in the child.
   np->trapframe->a0 = 0;
@@ -437,27 +467,38 @@ scheduler(void)
     intr_on();
     intr_off();
 
-    int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
+    int total = 0;
+    for(p = proc; p < &proc[NPROC]; p++){
       acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
-        found = 1;
-      }
+      if(p->state == RUNNABLE)
+        total += p->tickets;
       release(&p->lock);
     }
-    if(found == 0) {
+
+    if(total == 0) {
       // nothing to run; stop running on this core until an interrupt.
       asm volatile("wfi");
+      continue;
+    }
+
+    int winning = rand() % total;
+
+    int sum = 0;
+    for(p = proc; p < &proc[NPROC]; p++){
+      acquire(&p->lock);
+      if(p->state == RUNNABLE){
+        sum += p->tickets;
+        if(sum > winning){
+          p->state = RUNNING;
+          p->rounds++;
+          c->proc = p;
+          swtch(&c->context, &p->context);
+          c->proc = 0;
+          release(&p->lock);
+          break;
+        }
+      }
+      release(&p->lock);
     }
   }
 }
@@ -684,7 +725,7 @@ procdump(void)
       state = states[p->state];
     else
       state = "???";
-    printf("%d %s %s", p->pid, state, p->name);
+    printf("%d %s %s tickets=%d rounds=%d", p->pid, state, p->name, p->tickets, p->rounds);
     printf("\n");
   }
 }
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..b64918f 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -91,6 +91,8 @@ struct proc {
   int killed;                  // If non-zero, have been killed
   int xstate;                  // Exit status to be returned to parent's wait
   int pid;                     // Process ID
+  int tickets;
+  int rounds;
 
   // wait_lock must be held when using this:
   struct proc *parent;         // Parent process
diff --git a/kernel/syscall.c b/kernel/syscall.c
index 076d965..e8daf8c 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -101,6 +101,7 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_settickets(void);
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +127,7 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_settickets] sys_settickets,
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index 3dd926d..f085330 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,4 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_settickets 22
\ No newline at end of file
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 419e727..13f8ea9 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -7,6 +7,20 @@
 #include "proc.h"
 #include "vm.h"
 
+uint64
+sys_settickets(void)
+{
+  int n;
+
+  argint(0, &n); 
+
+  if(n <= 0)
+    return -1;
+
+  myproc()->tickets = n;
+  return 0;
+}
+
 uint64
 sys_exit(void)
 {
diff --git a/user/test_scheduler.c b/user/test_scheduler.c
new file mode 100644
index 0000000..04ba2b3
--- /dev/null
+++ b/user/test_scheduler.c
@@ -0,0 +1,19 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int
+main(int argc, char *argv[])
+{
+  if(argc != 2){
+    printf("Usage: test_scheduler tickets\n");
+    exit(1);
+  }
+
+  int tickets = atoi(argv[1]);
+  settickets(tickets);
+
+  while(1) {
+    ;
+  }
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index ac84de9..3d6ca08 100644
--- a/user/user.h
+++ b/user/user.h
@@ -24,6 +24,7 @@ int getpid(void);
 char* sys_sbrk(int,int);
 int pause(int);
 int uptime(void);
+int settickets(int);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index c5d4c3a..087dd83 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -42,3 +42,4 @@ entry("getpid");
 entry("sbrk");
 entry("pause");
 entry("uptime");
+entry("settickets");
